diff -Naur old/README.MacFUSE new/README.MacFUSE
--- old/README.MacFUSE	1969-12-31 16:00:00.000000000 -0800
+++ new/README.MacFUSE	2007-02-10 19:21:38.000000000 -0800
@@ -0,0 +1,7 @@
+MacFUSE Project Page:
+
+http://code.google.com/p/macfuse/
+
+Making a Universal build:
+
+CFLAGS="-D__FreeBSD__=10 -D_POSIX_C_SOURCE=200112L -O -g -arch i386 -arch ppc -isysroot /Developer/SDKs/MacOSX10.4u.sdk" LDFLAGS="-arch i386 -arch ppc" ./configure --prefix=/usr/local --disable-dependency-tracking
diff -Naur old/configure new/configure
--- old/configure	2007-02-04 06:30:48.000000000 -0800
+++ new/configure	2007-02-10 19:21:38.000000000 -0800
@@ -11284,7 +11284,8 @@
 fi
 
 
-libfuse_libs=-pthread
+#libfuse_libs=-pthread
+libfuse_libs=
 LIBS=
 { echo "$as_me:$LINENO: checking for library containing clock_gettime" >&5
 echo $ECHO_N "checking for library containing clock_gettime... $ECHO_C" >&6; }
diff -Naur old/example/fusexmp.c new/example/fusexmp.c
--- old/example/fusexmp.c	2006-12-09 10:51:13.000000000 -0800
+++ new/example/fusexmp.c	2007-02-10 19:21:38.000000000 -0800
@@ -45,8 +45,25 @@
     int res;
 
     res = access(path, mask);
-    if (res == -1)
+    if (res == -1) {
+#if (__FreeBSD__ >= 10)
+        if (errno == ENOENT) {
+            int next_res;
+            struct stat sb;
+            next_res = lstat(path, &sb);
+            if (next_res == 0 && S_ISLNK(sb.st_mode) &&
+                ((sb.st_uid == getuid()) || (geteuid() == 0))) {
+                /*
+                 * The target of the symlink doesn't exist, but the
+                 * symlink itself does. A lame cheap fix here.
+                 */
+                return 0;
+            }
+            return -ENOENT;
+        }
+#endif
         return -errno;
+    }  /* if (res == -1) */
 
     return 0;
 }
@@ -314,7 +331,11 @@
 static int xmp_setxattr(const char *path, const char *name, const char *value,
                         size_t size, int flags)
 {
+#if (__FreeBSD__ >= 10)
+    int res = setxattr(path, name, value, size, 0, flags);
+#else
     int res = lsetxattr(path, name, value, size, flags);
+#endif
     if (res == -1)
         return -errno;
     return 0;
@@ -323,7 +344,11 @@
 static int xmp_getxattr(const char *path, const char *name, char *value,
                     size_t size)
 {
+#if (__FreeBSD__ >= 10)
+    int res = getxattr(path, name, value, size, 0, 0);
+#else
     int res = lgetxattr(path, name, value, size);
+#endif
     if (res == -1)
         return -errno;
     return res;
@@ -331,7 +356,11 @@
 
 static int xmp_listxattr(const char *path, char *list, size_t size)
 {
+#if (__FreeBSD__ >= 10)
+    int res = listxattr(path, list, size, 0);
+#else
     int res = llistxattr(path, list, size);
+#endif
     if (res == -1)
         return -errno;
     return res;
@@ -339,7 +368,11 @@
 
 static int xmp_removexattr(const char *path, const char *name)
 {
+#if (__FreeBSD__ >= 10)
+    int res = removexattr(path, name, 0);
+#else
     int res = lremovexattr(path, name);
+#endif
     if (res == -1)
         return -errno;
     return 0;
diff -Naur old/example/fusexmp_fh.c new/example/fusexmp_fh.c
--- old/example/fusexmp_fh.c	2006-12-09 10:51:13.000000000 -0800
+++ new/example/fusexmp_fh.c	2007-02-10 19:21:38.000000000 -0800
@@ -32,8 +32,9 @@
     int res;
 
     res = lstat(path, stbuf);
-    if (res == -1)
+    if (res == -1) {
         return -errno;
+    }
 
     return 0;
 }
@@ -57,8 +58,25 @@
     int res;
 
     res = access(path, mask);
-    if (res == -1)
+    if (res == -1) {
+#if (__FreeBSD__ >= 10)
+        if (errno == ENOENT) {
+            int next_res;
+            struct stat sb;
+            next_res = lstat(path, &sb);
+            if (next_res == 0 && S_ISLNK(sb.st_mode) &&
+                ((sb.st_uid == getuid()) || (geteuid() == 0))) {
+                /*
+                 * The target of the symlink doesn't exist, but the
+                 * symlink itself does. A lame cheap fix here.
+                 */
+                return 0;
+            }
+            return -ENOENT;
+        }
+#endif
         return -errno;
+    }  /* if (res == -1) */
 
     return 0;
 }
@@ -266,7 +284,11 @@
 {
     int fd;
 
+#if (__FreeBSD__ >= 10)
+    fd = open(path, fi->flags | O_CREAT, mode);
+#else
     fd = open(path, fi->flags, mode);
+#endif
     if (fd == -1)
         return -errno;
 
@@ -373,7 +395,11 @@
 static int xmp_setxattr(const char *path, const char *name, const char *value,
                         size_t size, int flags)
 {
+#if (__FreeBSD__ >= 10)
+    int res = setxattr(path, name, value, size, 0, flags);
+#else
     int res = lsetxattr(path, name, value, size, flags);
+#endif
     if (res == -1)
         return -errno;
     return 0;
@@ -382,7 +408,11 @@
 static int xmp_getxattr(const char *path, const char *name, char *value,
                     size_t size)
 {
+#if (__FreeBSD__ >= 10)
+    int res = getxattr(path, name, value, size, 0, 0);
+#else
     int res = lgetxattr(path, name, value, size);
+#endif
     if (res == -1)
         return -errno;
     return res;
@@ -390,7 +420,11 @@
 
 static int xmp_listxattr(const char *path, char *list, size_t size)
 {
+#if (__FreeBSD__ >= 10)
+    int res = listxattr(path, list, size, 0);
+#else
     int res = llistxattr(path, list, size);
+#endif
     if (res == -1)
         return -errno;
     return res;
@@ -398,7 +432,11 @@
 
 static int xmp_removexattr(const char *path, const char *name)
 {
+#if (__FreeBSD__ >= 10)
+    int res = removexattr(path, name, 0);
+#else
     int res = lremovexattr(path, name);
+#endif
     if (res == -1)
         return -errno;
     return 0;
@@ -414,7 +452,20 @@
                        sizeof(fi->lock_owner));
 }
 
+void *
+xmp_init(struct fuse_conn_info *conn)
+{
+    return NULL;
+}
+
+void
+xmp_destroy(void *userdata)
+{
+}
+
 static struct fuse_operations xmp_oper = {
+    .init       = xmp_init,
+    .destroy    = xmp_destroy,
     .getattr	= xmp_getattr,
     .fgetattr	= xmp_fgetattr,
     .access	= xmp_access,
diff -Naur old/fuse.pc.in new/fuse.pc.in
--- old/fuse.pc.in	2006-12-09 10:52:22.000000000 -0800
+++ new/fuse.pc.in	2007-02-10 19:21:38.000000000 -0800
@@ -4,7 +4,7 @@
 includedir=@includedir@
 
 Name: fuse
-Description: Filesystem in Userspace
+Description: File System in User Space (MacFUSE)
 Version: @VERSION@
 Libs: -L${libdir} -lfuse @libfuse_libs@
-Cflags: -I${includedir}/fuse -D_FILE_OFFSET_BITS=64
+Cflags: -I${includedir}/fuse -D__FreeBSD__=10 -D_FILE_OFFSET_BITS=64
diff -Naur old/include/fuse.h new/include/fuse.h
--- old/include/fuse.h	2007-01-31 12:15:26.000000000 -0800
+++ new/include/fuse.h	2007-02-10 19:21:38.000000000 -0800
@@ -662,4 +662,8 @@
 }
 #endif
 
+#if (__FreeBSD__ >= 10)
+#define __asm__(x)
+#endif
+
 #endif /* _FUSE_H_ */
diff -Naur old/lib/Makefile.in new/lib/Makefile.in
--- old/lib/Makefile.in	2007-02-04 06:31:02.000000000 -0800
+++ new/lib/Makefile.in	2007-02-10 19:21:38.000000000 -0800
@@ -60,8 +60,12 @@
 	fuse_loop.c fuse_loop_mt.c fuse_lowlevel.c fuse_misc.h \
 	fuse_mt.c fuse_opt.c fuse_session.c fuse_signals.c helper.c \
 	mount.c mount_bsd.c
+ifeq (@host_vendor@, apple)
+am__objects_1 = mount_bsd.lo
+else
 @BSD_FALSE@am__objects_1 = mount.lo
 @BSD_TRUE@am__objects_1 = mount_bsd.lo
+endif
 am_libfuse_la_OBJECTS = fuse.lo fuse_kern_chan.lo fuse_loop.lo \
 	fuse_loop_mt.lo fuse_lowlevel.lo fuse_mt.lo fuse_opt.lo \
 	fuse_session.lo fuse_signals.lo helper.lo $(am__objects_1)
@@ -202,8 +206,12 @@
  -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DFUSE_USE_VERSION=26
 
 lib_LTLIBRARIES = libfuse.la libulockmgr.la
+ifeq (@host_vendor@, apple)
+mount_source = mount_bsd.c
+else
 @BSD_FALSE@mount_source = mount.c
 @BSD_TRUE@mount_source = mount_bsd.c
+endif
 libfuse_la_SOURCES = \
 	fuse.c			\
 	fuse_i.h		\
@@ -219,8 +227,12 @@
 	helper.c		\
 	$(mount_source)
 
+ifeq ($(host_vendor), apple)
+libfuse_la_LDFLAGS = @libfuse_libs@
+else
 libfuse_la_LDFLAGS = @libfuse_libs@ -version-number 2:6:3 \
 	-Wl,--version-script,$(srcdir)/fuse_versionscript
+endif
 
 libulockmgr_la_SOURCES = ulockmgr.c
 libulockmgr_la_LDFLAGS = -version-number 1:0:1
diff -Naur old/lib/fuse.c new/lib/fuse.c
--- old/lib/fuse.c	2007-02-02 03:57:48.000000000 -0800
+++ new/lib/fuse.c	2007-02-10 19:21:38.000000000 -0800
@@ -734,6 +734,16 @@
 
 static void curr_time(struct timespec *now)
 {
+#if (__FreeBSD__ >= 10)
+#define MACFUSE_TIMEVAL_TO_TIMESPEC(tv, ts) {                           \
+        (ts)->tv_sec = (tv)->tv_sec;                                    \
+        (ts)->tv_nsec = (tv)->tv_usec * 1000;                           \
+    }
+    struct timeval tp;
+    gettimeofday(&tp, NULL);
+    /* XXX: TBD: We are losing resolution here. */
+    MACFUSE_TIMEVAL_TO_TIMESPEC(&tp, now);
+#else
     static clockid_t clockid = CLOCK_MONOTONIC;
     int res = clock_gettime(clockid, now);
     if (res == -1 && errno == EINVAL) {
@@ -744,6 +754,7 @@
         perror("fuse: clock_gettime");
         abort();
     }
+#endif
 }
 
 static void update_stat(struct node *node, const struct stat *stbuf)
diff -Naur old/lib/fuse_session.c new/lib/fuse_session.c
--- old/lib/fuse_session.c	2006-12-09 10:51:14.000000000 -0800
+++ new/lib/fuse_session.c	2007-02-10 19:21:38.000000000 -0800
@@ -15,6 +15,9 @@
 #include <string.h>
 #include <assert.h>
 #include <errno.h>
+#if (__FreeBSD__ >= 10)
+#include <sys/param.h>
+#endif
 
 struct fuse_session {
     struct fuse_session_ops op;
@@ -24,6 +27,10 @@
     volatile int exited;
 
     struct fuse_chan *ch;
+
+#if (__FreeBSD__ >= 10)
+    char mntonname[MAXPATHLEN];
+#endif
 };
 
 struct fuse_chan {
@@ -40,6 +47,30 @@
     int compat;
 };
 
+#if (__FreeBSD__ >= 10)
+char *fuse_session_get_mntonname(struct fuse_session *se)
+{
+    if (se) {
+        return se->mntonname;
+    }
+
+    return NULL;
+}
+
+void fuse_session_set_mntonname(struct fuse_session *se, char *mntonname)
+{
+    if (!se || !mntonname) {
+        return;
+    }
+
+    if (!realpath(mntonname, se->mntonname)) {
+        memset(se->mntonname, 0, MAXPATHLEN);
+    }
+
+    return;
+}
+#endif
+
 struct fuse_session *fuse_session_new(struct fuse_session_ops *op, void *data)
 {
     struct fuse_session *se = (struct fuse_session *) malloc(sizeof(*se));
diff -Naur old/lib/fuse_signals.c new/lib/fuse_signals.c
--- old/lib/fuse_signals.c	2006-12-09 10:51:14.000000000 -0800
+++ new/lib/fuse_signals.c	2007-02-10 19:21:38.000000000 -0800
@@ -13,12 +13,34 @@
 #include <signal.h>
 
 static struct fuse_session *fuse_instance;
+#if (__FreeBSD__ >= 10)
+extern char *fuse_session_get_mntonname(struct fuse_session *se);
+
+#include <unistd.h>
+#endif
 
 static void exit_handler(int sig)
 {
     (void) sig;
+#if (__FreeBSD__ >= 10)
+    if (fuse_instance && !fuse_session_exited(fuse_instance)) {
+        int fd;
+        pid_t pid;
+
+        fd = fuse_chan_fd(fuse_session_next_chan(fuse_instance, NULL));
+        pid = fork();
+        if (pid == 0) { /* child */
+             char *mntonname = fuse_session_get_mntonname(fuse_instance);
+             fcntl(fd, F_SETFD, 1); /* close-on-exec */
+             execl("/sbin/umount", "/sbin/umount", mntonname, NULL);
+        } else {
+            /* We do nothing in the parent. */
+        }
+    }
+#else
     if (fuse_instance)
         fuse_session_exit(fuse_instance);
+#endif
 }
 
 static int set_one_signal_handler(int sig, void (*handler)(int))
diff -Naur old/lib/helper.c new/lib/helper.c
--- old/lib/helper.c	2006-12-09 10:52:22.000000000 -0800
+++ new/lib/helper.c	2007-02-10 19:21:38.000000000 -0800
@@ -264,6 +264,10 @@
     if (fd)
         *fd = fuse_chan_fd(ch);
 
+#if (__FreeBSD__ >= 10)
+    fuse_session_set_mntonname(fuse_get_session(fuse), *mountpoint);
+#endif
+
     return fuse;
 
  err_unmount:
diff -Naur old/lib/mount_bsd.c new/lib/mount_bsd.c
--- old/lib/mount_bsd.c	2006-12-09 10:51:14.000000000 -0800
+++ new/lib/mount_bsd.c	2007-02-10 19:22:44.000000000 -0800
@@ -6,6 +6,11 @@
     See the file COPYING.LIB.
 */
 
+#if (__FreeBSD__ >= 10)
+#undef _POSIX_C_SOURCE
+#include <sys/types.h>
+#endif
+
 #include "fuse_i.h"
 #include "fuse_opt.h"
 
@@ -21,8 +26,69 @@
 #include <string.h>
 #include <paths.h>
 
+#if (__FreeBSD__ >= 10)
+#define FUSERMOUNT_PROG  "/System/Library/Filesystems/fusefs.fs/mount_fusefs"
+#define FUSE_DEV_TRUNK   "/dev/fuse"
+#define PRIVATE_LOAD_COMMAND "/Library/Extensions/fusefs.kext/Contents/Resources/load_fusefs"
+
+#include <sys/param.h>
+#include <sys/mount.h>
+#include <AssertMacros.h>
+
+static const char *MacFUSE = "MacFUSE version 0.2.0, " __DATE__ ", " __TIME__;
+
+static int
+checkloadable(void)
+{
+    int ret;
+    struct vfsconf vfc;
+    
+    ret = getvfsbyname("fusefs", &vfc);
+    
+    return ret;
+}   
+
+int
+loadkmod()
+{
+    int result = -1;
+    int pid, terminated_pid;
+    union wait status;
+
+    pid = fork();
+
+    if (pid == 0) {
+        result = execl(PRIVATE_LOAD_COMMAND, PRIVATE_LOAD_COMMAND, NULL);
+        
+        /* exec failed */
+        goto Return;
+    }
+
+    require_action(pid != -1, Return, result = errno);
+
+    while ((terminated_pid = wait4(pid, (int *)&status, 0, NULL)) < 0) {
+        /* retry if EINTR, else break out with error */
+        if ( errno != EINTR ) {
+            break;
+        }
+    }
+
+    if ((terminated_pid == pid) && (WIFEXITED(status))) {
+        result = WEXITSTATUS(status);
+    } else {
+        result = -1;
+    }
+
+Return:
+    check_noerr_string(result, strerror(errno));
+    
+    return result;
+}
+
+#else
 #define FUSERMOUNT_PROG         "mount_fusefs"
 #define FUSE_DEV_TRUNK          "/dev/fuse"
+#endif
 
 enum {
     KEY_ALLOW_ROOT,
@@ -108,11 +174,35 @@
     FUSE_OPT_KEY("noprivate",           KEY_KERN),
     FUSE_OPT_KEY("noneglect_shares",    KEY_KERN),
     FUSE_OPT_KEY("nopush_symlinks_in",  KEY_KERN),
+#if (__FreeBSD__ >= 10)
+    /* Mac OS X options */
+    FUSE_OPT_KEY("blocksize=",          KEY_KERN),
+    FUSE_OPT_KEY("daemon_timeout=",     KEY_KERN),
+    FUSE_OPT_KEY("extended_security",   KEY_KERN),
+    FUSE_OPT_KEY("fsid=",               KEY_KERN),
+    FUSE_OPT_KEY("fsname=",             KEY_KERN),
+    FUSE_OPT_KEY("init_timeout=",       KEY_KERN),
+    FUSE_OPT_KEY("iosize=",             KEY_KERN),
+    FUSE_OPT_KEY("jail_symlinks",       KEY_KERN),
+    FUSE_OPT_KEY("noattrcache",         KEY_KERN),
+    FUSE_OPT_KEY("noauthopaque",        KEY_KERN),
+    FUSE_OPT_KEY("noauthopaqueaccess",  KEY_KERN),
+    FUSE_OPT_KEY("nobrowse",            KEY_KERN),
+    FUSE_OPT_KEY("nolocalcaches",       KEY_KERN),
+    FUSE_OPT_KEY("noreadahead",         KEY_KERN),
+    FUSE_OPT_KEY("nosyncwrites",        KEY_KERN),
+    FUSE_OPT_KEY("noubc",               KEY_KERN),
+    FUSE_OPT_KEY("novncache",           KEY_KERN),
+    FUSE_OPT_KEY("ping_diskarb",        KEY_KERN),
+    FUSE_OPT_KEY("subtype=",            KEY_KERN),
+    FUSE_OPT_KEY("volname=",            KEY_KERN),
+#else
     /* Linux specific mount options, but let just the mount util handle them */
     FUSE_OPT_KEY("fsname=",             KEY_KERN),
     FUSE_OPT_KEY("nonempty",            KEY_KERN),
     FUSE_OPT_KEY("large_read",          KEY_KERN),
     FUSE_OPT_KEY("max_read=",           KEY_KERN),
+#endif
     FUSE_OPT_END
 };
 
@@ -199,8 +289,21 @@
     if (rv)
         return;
 
+#if (__FreeBSD__ >= 10)
+    {
+        int ret;
+        char *rp = NULL;
+        char resolved_path[PATH_MAX];
+
+        rp = realpath(mountpoint, resolved_path);
+        if (rp) {
+            ret = unmount(resolved_path, 0);
+        }
+    }
+#else
     asprintf(&umount_cmd, "/sbin/umount %s", dev);
     system(umount_cmd);
+#endif
 }
 
 void fuse_kern_unmount(const char *mountpoint, int fd)
@@ -222,13 +325,29 @@
     if (*ep != '\0')
         return;
 
+#if (__FreeBSD__ >= 10)
+    {
+        int ret;
+        char *rp = NULL;
+        char resolved_path[PATH_MAX];
+
+        rp = realpath(mountpoint, resolved_path);
+        if (rp) {
+            ret = unmount(resolved_path, 0);
+        }
+    }
+#else
     asprintf(&umount_cmd, "/sbin/umount " _PATH_DEV "%s", dev);
     system(umount_cmd);
+#endif
 }
 
 /* Check if kernel is doing init in background */
 static int init_backgrounded(void)
 {
+#if (__FreeBSD__ >= 10)
+    return 0;
+#else
     int ibg, len;
 
     len = sizeof(ibg);
@@ -237,6 +356,7 @@
         return 0;
 
     return ibg;
+#endif
 }
 
 
@@ -247,6 +367,15 @@
     char *fdnam, *dev;
     int pid;
 
+#if (__FreeBSD__ >= 10)
+    if (checkloadable()) {
+        if (loadkmod()) {
+            fprintf(stderr, "fusefs file system is not available");
+            return -1;
+        }
+    }
+#endif
+
     fdnam = getenv("FUSE_DEV_FD");
 
     if (fdnam) {
@@ -267,12 +396,29 @@
 
     dev = getenv("FUSE_DEV_NAME");
 
-    if (! dev)
-	dev = FUSE_DEV_TRUNK;
-
-    if ((fd = open(dev, O_RDWR)) < 0) {
-        perror("fuse: failed to open fuse device");
-        return -1;
+    if (dev) {
+        if ((fd = open(dev, O_RDWR)) < 0) {
+            perror("fuse: failed to open fuse device");
+            return -1;
+        }
+    } else {
+#define NFUSEDEVICE 16
+        int r, devidx = -1;
+        char devpath[MAXPATHLEN];
+
+        for (r = 0; r < NFUSEDEVICE; r++) {
+            snprintf(devpath, MAXPATHLEN - 1, "/dev/fuse%d", r);
+            fd = open(devpath, O_RDWR);
+            if (fd >= 0) {
+                dev = devpath;
+                devidx = r;
+                break;
+            }
+        }
+        if (devidx == -1) {
+            perror("fuse: failed to open fuse device");
+            return -1;
+        }
     }
 
 mount:
@@ -322,11 +468,31 @@
             perror("fuse: failed to exec mount program");
             exit(1);
         }
+#if (__FreeBSD__ >= 10)
+        else {
+            int status;
+            waitpid(pid, &status, 0);
+            if (WIFEXITED(status) && (WEXITSTATUS(status) != 0)) {
+                exit(WEXITSTATUS(status));
+            }
+        }
+#endif
 
         exit(0);
     }
 
+#if (__FreeBSD__ >= 10)
+    {
+        int status;
+
+        waitpid(pid, &status, 0);
+        if (WIFEXITED(status) && (WEXITSTATUS(status) != 0)) {
+            exit(WEXITSTATUS(status));
+        }
+    }
+#else
     waitpid(pid, NULL, 0);
+#endif
 
 out:
     return fd;
diff -Naur old/lib/ulockmgr.c new/lib/ulockmgr.c
--- old/lib/ulockmgr.c	2007-01-02 09:10:33.000000000 -0800
+++ new/lib/ulockmgr.c	2007-02-10 19:21:38.000000000 -0800
@@ -18,9 +18,19 @@
 #include <assert.h>
 #include <signal.h>
 #include <sys/stat.h>
+#if (__FreeBSD__ >= 10)
+#undef _POSIX_C_SOURCE
 #include <sys/socket.h>
+#define _POSIX_C_SOURCE 200112L
+#else
+#include <sys/socket.h>
+#endif
 #include <sys/wait.h>
 
+#if (__FreeBSD__ >= 10)
+#define MSG_NOSIGNAL 0
+#endif
+
 struct message {
     unsigned intr : 1;
     unsigned nofd : 1;
@@ -125,6 +135,13 @@
         return -1;
     }
     ulockmgr_cfd = sv[1];
+#if (__FreeBSD__ >= 10)
+    {
+        int on = 1;
+        res = setsockopt(ulockmgr_cfd, SOL_SOCKET, SO_NOSIGPIPE, (void *)&on,
+                         sizeof(on));
+    }
+#endif
     return 0;
 }
 
